add trait RDFLoader, using both content negotiation or URI extension to guess the actual RDF syntax

jmvanel opened the pull request:
with a Jena implementation based on

``` scala
org.apache.jena.riot.RDFDataMgr.loadModel(uri)
```

and a test for Jena.
See 
https://jena.apache.org/documentation/javadoc/arq/org/apache/jena/riot/RDFDataMgr.html#loadModel(java.lang.String)

I don't think Sesame provides a similar facility, but it has some useful helpers , but no smart reader from an URL : 
http://rdf4j.org/sesame/2.7/docs/users.docbook?view#chapter-rio4

Anyway, the PR is useful as is for Jena users .

https://jena.apache.org/documentation/javadoc/arq/org/apache/jena/riot/RDFDataMgr.html#read(com.hp.hpl.jena.graph.Graph,%20java.lang.String)

jmvanel committed:
Add trait RDFLoader, an RDF loader that uses both content negotiation or URI extension to guess the actual RDF syntax, with Jena (native) & Sesame implementations

add jsonldReader in JenaModule
FromLiteral.scala : add implicit def LongFromLiteral , FloatFromLiteral

bblfish commented on jena/src/main/scala/org/w3/banana/jena/io/JenaRDFReader.scala:
```diff
@@ -62,10 +62,22 @@ object JenaRDFReader {
+     *  (content negotiation or extension).
+     */
+    def readAnyRDFSyntax(uri: String): Try[Jena#Graph] = Try {
+      val model = RDFDataMgr.loadModel(uri)
```
One problem lies here. Using the built in RDF loaders from Jena or sesame to do remote fetching of documents is like going back to the stone ages ( I exaggerate of course ). All IO should be in a separate package, and not tied into the RDF core, as that can be done so much better with Scala, using frameworks such as aka.http .

I don't think we have consensus on the IO yet.

betehess commented on rdf/common/src/main/scala/org/w3/banana/io/AnyRDFReader.scala:
```diff
@@ -0,0 +1,12 @@
+  
+	/** Read triples from the given location. The syntax is determined from input source URI
+   *  (content negotiation or extension). */
+   def readAnyRDFSyntax( uri:String) : M[Rdf#Graph]
```
As mentioned by @bblfish that stuff should go in `io`, at least for now.

I personally don't like the name `AnyRDFReader`. It has nothing to do with `RDFReader`. Also `uri` should at least be `url`. Let's get this type right as well (I know it's not the case for other abstractions yet): either `java.net.URL` (it's ok for scala-js) or `Rdf#URI`.

Here is a proposal:

``` scala
trait RDFLoader[Rdf <: RDF, M[_]] {
   def load(url: java.net.URL) : M[Rdf#Graph]
}
```

jmvanel replied:
Why not ?  But this RDFReader with 2 parameters would be confusing with the original one with 3 parameters.

betehess replied:
What `RDFReader`?

jmvanel replied:
Sorry I misread your suggestion.

jmvanel replied:
So a new SBT subproject , like io/rdfLoaders  ?

betehess replied:
Ah, I just realized that we were talking about different things. I read "separate package" from @bblfish's comment and understood it as "separate module". IMO, under `https://github.com/w3c/banana-rdf/tree/series/0.8.x/rdf/common/src/main/scala/org/w3/banana/io` is just fine for the abstractions. The implementations would live in jena and sesame, as expected. So as far as I am concerned, you're already fine.

betehess commented on jena/src/test/scala/org/w3/banana/jena/TestAnyRDFReader.scala:
```diff
@@ -0,0 +1,50 @@
+trait TestAnyRDFReader extends AnyRDFReaderModule with RDFOpsModule {
+  import ops._
+
+  val rdfxmlDocURL = "http://eulergui.sourceforge.net/doap_eulergui.rdf"
```
Please no tests that depend on external resources. It's too easy to get failing tests because of that.

betehess commented on jena/src/main/scala/org/w3/banana/jena/io/JenaRDFReader.scala:
```diff
@@ -62,10 +62,22 @@ object JenaRDFReader {
     }
   }
 
+  def makeAnyRDFReader = new AnyRDFReader[Jena, Try] {
```
Make it a real function, add `()`. And make sure you call it right in the module.

betehess commented:
General comment: please propose an implementation for sesame as well.

InTheNow commented on sesame/src/main/scala/org/w3/banana/sesame/io/SesameRDFReader.scala:
```diff
@@ -75,3 +75,62 @@ class SesameJSONLDReader(implicit val ops: RDFOps[Sesame]) extends AbstractSesam
+      val connection = url.openConnection()
+      connection match {
+        case connection: HttpURLConnection =>
+        val acceptHeaderTurtlePriority =
```
I wonder if this could be better achieved with a shapless HList (or similar) passed in as the third type parameter to RDFLoader? As well as the resulting signature being in common with the other readers, it would remove the hard coding of the priority and would be open for new reader types without modification?

Interestingly, that implementation would be richer that than the standard jena one.

jmvanel replied:
I don't know  shapeless and, sorry, no time for learning that at the moment.

InTheNow commented on sesame/src/main/scala/org/w3/banana/sesame/SesameModule.scala:
```diff
@@ -2,7 +2,7 @@ package org.w3.banana.sesame
 import org.openrdf.repository.RepositoryConnection
 import org.w3.banana._
-import org.w3.banana.io.{SparqlQueryResultsReader, _}
+import org.w3.banana.io.{ SparqlQueryResultsReader, _ }
```
b-rdf uses both styles but I believe the original is the generally preferred style (note: not __my__ preference)

jmvanel replied:
DONE

InTheNow commented on sesame/src/main/scala/org/w3/banana/sesame/io/SesameRDFReader.scala:
```diff
@@ -5,7 +5,7 @@ import java.util.LinkedList
 import com.github.jsonldjava.sesame.SesameJSONLDParser
 import org.openrdf.model._
-import org.openrdf.model.impl.{LinkedHashModel, LiteralImpl, StatementImpl}
+import org.openrdf.model.impl.{ LinkedHashModel, LiteralImpl, StatementImpl }
```
style

jmvanel replied:
I see in CONTRIBUTING.md that the style is now with spaces :) .

InTheNow commented on jena/src/main/scala/org/w3/banana/jena/io/JenaRDFReader.scala:
```diff
@@ -1,12 +1,12 @@
 
-import com.hp.hpl.jena.graph.{Node => JenaNode, Triple => JenaTriple, _}
-import com.hp.hpl.jena.rdf.model.{RDFReader => _}
+import com.hp.hpl.jena.graph.{ Node => JenaNode, Triple => JenaTriple, _ }
```
extra spaces, also two others

jmvanel replied:
DONE as in CONTRIBUTING.md

InTheNow commented on jena/src/main/scala/org/w3/banana/jena/io/JenaRDFReader.scala:
```diff
@@ -62,10 +62,21 @@ object JenaRDFReader {
   }
 
+  def makeAnyRDFReader() = new RDFLoader[Jena, Try] {
+    /**
```
Incorrect [ScalaDoc style](http://docs.scala-lang.org/style/scaladoc.html):

```
  /** Read triples from the given location. 
    *
    * The syntax is determined from input source URI 
    * (content negotiation or extension).
    */
```

jmvanel replied:
DONE

InTheNow commented on jena/src/main/scala/org/w3/banana/jena/io/JenaRDFReader.scala:
```diff
@@ -23,7 +23,7 @@ final class TripleSink(implicit ops: JenaOps) extends StreamRDF {
   def triple(triple: JenaTriple): Unit = {
     def isXsdString(node: JenaNode): Boolean =
       node.isLiteral &&
```
Although not originating from this PR, the body must be enclosed in braces as it is not a single line.

jmvanel replied:
DONE

jmvanel replied:
NOTE: this formatting rule is not in CONTRIBUTING.md

InTheNow commented on jena/src/main/scala/org/w3/banana/jena/io/JenaRDFReader.scala:
```diff
@@ -62,10 +62,21 @@ object JenaRDFReader {
+     * Read triples from the given location. The syntax is determined from input source URI
+     *  (content negotiation or extension).
+     */
+    def load(url: java.net.URL): Try[Jena#Graph] = {
```
move `Try` out of braces:
```
def load(url: java.net.URL): Try[Jena#Graph] = Try {
    RDFDataMgr.loadGraph(url.toString)
  }
```

jmvanel replied:
DONE

InTheNow commented on jena/src/test/scala/org/w3/banana/jena/JenaRDFLoader.scala:
```diff
@@ -0,0 +1,8 @@
+package org.w3.banana.jena
+
+import org.scalatest.FunSuite
```
For scalatest, wordspec is used, but this file should not be needed, rather add the test to [JenaTest.scala](https://github.com/w3c/banana-rdf/blob/series/0.8.x/jena/src/test/scala/org/w3/banana/jena/JenaTest.scala)

jmvanel replied:
DONE

InTheNow commented on rdf-test-suite/jvm/src/main/scala/org/w3/banana/io/RDFLoaderTestSuite.scala:
```diff
@@ -0,0 +1,45 @@
+package org.w3.banana.io
+
+import org.scalatest.FunSuite
```
use wordspec and general conventions used in other tests

jmvanel replied:
DONE

InTheNow commented on rdf/common/src/main/scala/org/w3/banana/RDFModule.scala:
```diff
@@ -78,6 +78,10 @@ trait JsonLDWriterModule extends RDFModule {
 }
 
+trait  RDFLoaderModule extends RDFModule {
+  implicit val rdfLoader: RDFLoader[Rdf, Try]
```
add blank lines before and after as per other implicit vals in file

jmvanel replied:
DONE,
but what are these blank lines for ?

InTheNow commented on rdf/common/src/main/scala/org/w3/banana/io/AnyRDFReader.scala:
```diff
@@ -0,0 +1,12 @@
+/** RDF reader. */
+trait RDFLoader[Rdf <: RDF, M[_]] {
+  
+	/** Read triples from the given location. The syntax is determined from input source URI
```
indentation and scaladoc style

jmvanel replied:
DONE

InTheNow commented on rdf/common/src/main/scala/org/w3/banana/io/AnyRDFReader.scala:
```diff
@@ -0,0 +1,12 @@
+
+
+/** RDF reader. */
+trait RDFLoader[Rdf <: RDF, M[_]] {
```
rename file to trait name

jmvanel replied:
DONE

InTheNow commented on rdf/common/src/main/scala/org/w3/banana/io/AnyRDFReader.scala:
```diff
@@ -0,0 +1,12 @@
+package io
+
+
+/** RDF reader. */
```
Comment does not match trait

InTheNow commented on sesame/src/main/scala/org/w3/banana/sesame/io/SesameRDFReader.scala:
```diff
@@ -75,3 +75,62 @@ class SesameJSONLDReader(implicit val ops: RDFOps[Sesame]) extends AbstractSesam
 
+
+
+class SesameRDFLoader(implicit val ops: RDFOps[Sesame]) extends RDFLoader[Sesame, Try] {
```
Should really go in its own file as its not a reader

jmvanel replied:
DONE

InTheNow commented on sesame/src/main/scala/org/w3/banana/sesame/io/SesameRDFReader.scala:
```diff
@@ -75,3 +75,62 @@ class SesameJSONLDReader(implicit val ops: RDFOps[Sesame]) extends AbstractSesam
+        }
+        case _ => parserForFileName
+    }
+    println("format: " + format)
```
What is the println for?

jmvanel replied:
was debug, commented out

InTheNow commented on sesame/src/test/scala/org/w3/banana/sesame/SesameRDFLoader.scala:
```diff
@@ -0,0 +1,8 @@
+import org.scalatest.FunSuite
+import org.w3.banana.io.RDFLoaderTestSuite
+
+class SesameRDFLoader extends FunSuite
```
should be in `SesameTest.scala`

jmvanel replied:
DONE

InTheNow commented:
A general comment, for all our PR's in fact, is that all the commits should be squashed before a final merge.

bblfish commented on sesame/src/main/scala/org/w3/banana/sesame/io/SesameRDFReader.scala:
```diff
@@ -75,3 +75,62 @@ class SesameJSONLDReader(implicit val ops: RDFOps[Sesame]) extends AbstractSesam
+      connection match {
+        case connection: HttpURLConnection =>
+        val acceptHeaderTurtlePriority =
+            "text/n3," +
```
text/n3 has to be lower at present, because there are very few parsers that actually parse that. In fact the ordering of these should depend on the parsers available. In any case neither Sesame nor Jena I think have real N3 parsers. Sadly.

jmvanel replied:
Alas, you're right,
I kept n3 but in last position.

antonkulaga commented:
as @bblfish merged my PR, could you rebase on top of the current master?

bblfish commented:
yes, it looks like there are conflicts that must be resolved....

jmvanel committed:
Merge branch 'series/0.8.x' into series/0.8.x

jmvanel committed:
use new Jena 3.4.0 API in JenaRDFReader in read(is: InputStream, base: String)

This should fix problem with relative URI like <file.ttl> , that were resolved as file://

jmvanel committed:
upgrade Jena, Fuseki, crossScalaVersions & build.properties to latest versions

jmvanel committed:
use new Jena 3.4.0 API in JenaRDFReader in read(reader: Reader, base: String)

- this removes a asInstanceOf[StringReader] and introduces a dependency to org.apache.commons.io.input.ReaderInputStream
- this removes deprecated calls to Jena

jmvanel committed:
FIX compile error: reference to RDFWriter is ambiguous

jmvanel committed:
Upgrade scalajs to 0.6.20

( anyway 0.6.13 was not compatible with Scala 11.11 )

bblfish commented:
On the whole I find the notion of an RDFLoader very shaky. It's problematic for many reasons, most of all that it is synchonous and blocking. So I see it can be useful to allow people to get going, but really it is terrible to do things this way nowadays. Also the mime types are going to be different in every framework used....

antonkulaga commented:
@bblfish you can revert if you want

antonkulaga commented:
the problem here is that there are many PRs that are ignored. Looks obvious that the lib needs new or more maintainers

bblfish commented:
I just studied the code today for the first time: I was too tired before writing EU proposals, and then recently working on my PhD thesis. (I have to give in my first year report soon)

So I had not completely come to a conclusion about this, because I see that a lot of effort has gone to this, and it should have been mentioned before if there was a problem.

jmvanel commented:
Henry, about the problem of RDFLoader being synchonous and blocking, if you could propose a better API for pulling triples out of an URL, I could try to implement it .
